packages/block-editor/src/components/block-list/block-invalid-warning.js

```diff
@@ -15,43 +14,90 @@ import Warning from '../warning';
 import BlockCompare from '../block-compare';
 import { store as blockEditorStore } from '../../store';
 
-export function BlockInvalidWarning( {
-	convertToHTML,
-	convertToBlocks,
-	convertToClassic,
-	attemptBlockRecovery,
-	block,
-} ) {
-	const hasHTMLBlock = !! getBlockType( 'core/html' );
-	const hasClassicBlock = !! getBlockType( 'core/freeform' );
-	const [ compare, setCompare ] = useState( false );
+const blockToBlocks = ( block ) =>
+	rawHandler( {
+		HTML: block.originalContent,
+	} );
+
+const recoverBlock = ( block ) =>
+	createBlock( block.name, block.attributes, block.innerBlocks );
+
+const blockToClassic = ( block ) =>
+	createBlock( 'core/freeform', {
+		content: block.originalContent,
+	} );
 
-	const onCompare = useCallback( () => setCompare( true ), [] );
+const blockToHTML = ( block ) =>
+	createBlock( 'core/html', {
+		content: block.originalContent,
+	} );
+
+export default function BlockInvalidWarning( { clientId } ) {
+	const { block, canInsertHTMLBlock, canInsertClassicBlock } = useSelect(
+		( select ) => {
+			const { canInsertBlockType, getBlock, getBlockRootClientId } =
+				select( blockEditorStore );
+
+			const rootClientId = getBlockRootClientId( clientId );
+
+			return {
+				block: getBlock( clientId ),
+				canInsertHTMLBlock: canInsertBlockType(
+					'core/html',
+					rootClientId
+				),
+				canInsertClassicBlock: canInsertBlockType(
+					'core/freeform',
+					rootClientId
+				),
+			};
+		},
+		[ clientId ]
+	);
+	const { replaceBlock } = useDispatch( blockEditorStore );
+
+	const [ compare, setCompare ] = useState( false );
 	const onCompareClose = useCallback( () => setCompare( false ), [] );
 
-	// We memo the array here to prevent the children components from being updated unexpectedly.
-	const hiddenActions = useMemo(
+	const attemptBlockRecovery = useCallback( () => {
+		replaceBlock( block.clientId, recoverBlock( block ) );
+	}, [ block, replaceBlock ] );
```

Comment:
Nit: There's no need for the `useCallback` for this function; we can inline it. The button component isn't memoized and will re-render alongside the parent.

Reply:
Yeah, I realise that. I did it for consistency, as otherwise the code looked ugly 😄